File: Python/other/Unique_number_III.py
# Unique_Number_III
# Given an array that contains all elements occurring 3 times, but one
# occurs only once. Find that unique element.


def unique_element(array, size):
    for i in range(size):
        j = 0
        while(j < size):
            if (i != j and array[i] == array[j]):
                break
            j += 1
        if (j == size):
            return array[i]

    return -1


print("Enter Elements:")
array = input().split()
size = len(array)
print("Unique element:")
print(unique_element(array, size))

'''
Sample I/O:
    Input:
    Enter Elements:
    1 1 1 4 2 2 2 3 3 3

    Output:
    Unique element:
    4

Time Complexity: O(size*size)
Space Complexity:O(1)
'''

File: C/other/UniqueNumberIII.c
#include<stdio.h>

main()
{
    int n;
    printf("Enter number of elements: ");
    scanf("%d",&n);
    int arr[n], i;
    printf("\nEnter %d elements: ", n);
    for(i = 0 ;i < n ;i++)
    {
        scanf("\n%d",&arr[i]);
    }
    int one = 0;
    int two = 0;
    int n_t, x;
    for(i = 0; i < n; i++)
    {
        x = arr[i];
        two |=one & x;
        one ^= x;
        n_t  =~(one & two);
        one &= n_t;
        two &= n_t;
    }
    printf("\nThe number that occurs only once is %d \n", one);
}

/*
Time complexity: O(n)
Space complexity: O(n)

Example 1:
Sample input:
Enter number of elements: 7
Enter 7 elements: 6 
2
5
2
2
6
6 
Output: The number that occurs only once is 5

Example 2:
Sample Input: 
Enter number of elements: 4
Enter 4 elements: 3 
3
1
3 
Output: The number that occurs only once is 1 
*/
